srk/src/polyhedronLatticeTiling.ml

```diff
@@ -0,0 +1,1486 @@
+
+  let local_project_polyreccone ~man ~max_dim_in_target =
+    let elim dim = dim > max_dim_in_target in
+    let abstract m plt =
+      let projected_polyhedron_generators =
+        CloseStrictIneq.polyhedral_level_set.abstract m plt
+        |> (LwCooper.real_local_project ~elim).abstract m
+        |> P.dd_of ~man (max_dim_in_target + 1)
+        |> DD.enum_generators
+      in
+      let reccone_generators =
+        CloseStrictIneq.round_assuming_no_ints.abstract m plt
+        |> (LwCooper.real_local_project ~elim).abstract m
+        |> P.dd_of ~man (max_dim_in_target + 1)
+        |> DD.enum_generators
+        |> BatEnum.filter (fun (kind, _) -> kind = `Ray || kind = `Line)
+      in
+      BatEnum.append projected_polyhedron_generators reccone_generators
+      |> DD.of_generators ~man (max_dim_in_target + 1)
+    in
+    LocalAbstraction.{
+      abstract
+    ; translate =
+        (fun m dim ->
+          if dim > max_dim_in_target then
+            failwith
+              (Format.asprintf
+                 "local_project_polyreccone: Dimension %d has been eliminated" dim)
+         else m dim)
+    }
+
+  let by_polyreccone_and_lplh ~man ~max_dim_in_target ~epsilon =
+    LocalAbstraction.join DD.join
+      (local_project_polyreccone ~man ~max_dim_in_target)
+      (local_project_local_hull ~man ~max_dim_in_target ~epsilon)
+
+end
+
+module ConvexHull : sig
+
+  (** Local abstraction algorithms for computing the closed convex hull of the image of
+      the set of models of an LRA, LIA, or LIRA formula under a linear map
+      defined by a set of terms.
+
+      More precisely,
+      if [alpha = convex_hull srk terms symbols],
+      [alpha m F] is the convex hull of { T(m): m |= F }, where
+      [T(m) = (terms[0](m), ..., terms[length(terms)-1](m))];
+      [t(m)] is the evaluation of term [t] on [m].
+
+      Formulas should only have linear real arithmetic terms,
+      i.e., no floor, mod, non-trivial multiplication and division, etc.
+   *)
+
+  (** All symbols should be of real type,
+      and all atoms are equalities or inequalities, i.e., no [is_int]'s.
+      Otherwise, it is an over-approximation of the result.
+   *)
+  val by_lp_assuming_real:
+    ?man: DD.closed Apron.Manager.t ->
+    'a context -> 'a arith_term array -> Symbol.Set.t ->
+    ('a formula, 'a Interpretation.interpretation, DD.closed DD.t, int -> QQ.t) LocalAbstraction.t
+
+  val by_full_project_assuming_real:
+    ?man:Polka.loose Polka.t Apron.Manager.t ->
+    'a context ->
+    'a arith_term array ->
+    Symbol.Set.t ->
+    ('a formula, 'a Interpretation.interpretation, DD.closed DD.t, int -> Q.t) LocalAbstraction.t
+
+  (** All symbols must be of integer type, and all terms must have integer coefficients.
+      Local-project-local-hull is (sound and) compact when these conditions hold.
+
+      For the latter condition, rounding assuming integer-valued variables is in general
```

Comment:
"terms must have integer coefficients" means the terms in the `'a arith_term array`?

Reply:
Yes. The example in the comment shows why; integrality constraints are not taken into account when we round.